onnxruntime/core/providers/cuda/tensor/transpose.cc

```diff
@@ -101,7 +101,42 @@ Status Transpose::DoTranspose(const cudaDeviceProp& prop,
   std::vector<int64_t> new_input_dims(input_dims.begin(), input_dims.end());
   std::vector<int64_t> new_output_dims(output_dims.begin(), output_dims.end());
 
-  for (auto i = rank - 1; i > 0; i--) {
+  // Remove all dims with value 1.
+  std::vector<bool> dims_to_remove(new_rank, false);
+  int input_pos = 0;
+  int output_pos = 0;
+  int perm_pos = 0;
+  for (int i = 0; i < new_rank; ++i) {
+    if (new_input_dims[i] != 1) {
+      new_input_dims[input_pos++] = new_input_dims[i];
+    } else {
+      dims_to_remove[i] = true;
+    }
+    if (new_output_dims[i] != 1) {
+      new_output_dims[output_pos++] = new_output_dims[i];
+    }
+  }
+  for (int i = 0; i < new_rank; ++i) {
+    if (!dims_to_remove[new_permutations[i]]) {
+      new_permutations[perm_pos++] = new_permutations[i];
+    }
+  }
+  for (int i = new_rank - 1; i >= 0; --i) {
+    if (dims_to_remove[i]) {
+      for (int j = 0; j < perm_pos; ++j) {
+        if (new_permutations[j] > static_cast<size_t>(i)) {
+          new_permutations[j] -= 1;
+        }
+      }
+    }
+  }
+  ORT_ENFORCE(input_pos == output_pos && input_pos == perm_pos);
+  new_rank = input_pos;
+  new_input_dims.resize(new_rank);
```

Comment:
>resize [](http://example.com/codeflow?start=17&length=6)

reserve() + push_back

Reply:
It modifies the same vector, so resize() is called at the end. Otherwise, we need to create a new vector and copy over all data, and need to modify all the following places that use the vector. The rank is normally very small, most of the time <=4, I am guessing it won't effect the perf a lot.